File: DDDSample.DomainVB/Aggregates/Customer/CustomerApplicationService.vb
Imports DDDSample

Public Class CustomerApplicationService
    Implements IApplicationService

    ReadOnly _eventStore As IEventStore

    Public Sub New(eventStore As IEventStore)
        If eventStore Is Nothing Then Throw New ArgumentNullException("eventStore")
        _eventStore = eventStore
    End Sub


    Public Sub Execute(command As ICommand) Implements IApplicationService.Execute
        [When](command)
    End Sub

    Public Sub [When](command As CreateCustomer)
        Update(command.Id, Sub(a) a.Create(command.Id, command.CustomerName))
    End Sub
    Public Sub Update(customerID As CustomerId, execute As Action(Of CustomerAggregate))
        While True
            Dim eventStream = _eventStore.LoadEventStream(id:=customerID)
            Dim customer As New CustomerAggregate(eventStream.Events)
            execute(customer)

            Try
                _eventStore.AppendEventsToStream(id:=customerID,
                                           version:=eventStream.StreamVersion,
                                           events:=customer.Changes)

            Catch ex As OptimisticConcurrencyException
                For Each customerEvent In customer.Changes
                    For Each actutalEvent In ex.ActualEvents
                        If ConflictsWith(customerEvent, actutalEvent) Then
                            Dim msg = String.Format("Conflict between {0} and {1}", customerEvent, actutalEvent)
                            Throw New RealConcurrencyException(msg, ex)
                        End If
                    Next
                Next
            End Try
        End While
    End Sub
    Public Function ConflictsWith(x As IEvent, y As IEvent)
        Return x.GetType.Equals(y.GetType)
    End Function
End Class

File: DDDSample.Domain/Aggregates/Customer/CustomerApplicationService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace DDDSample
{
    public class CustomerApplicationService : IApplicationService
    {
        readonly IEventStore _eventStore;
        readonly IDispatcher _dispatcher;

        public CustomerApplicationService(IEventStore eventStore,
                                          IDispatcher dispatcher)
        {
            if (eventStore == null) throw new ArgumentNullException("eventStore");
            if (dispatcher == null) throw new ArgumentNullException("dispatcher");
            _eventStore = eventStore;
            _dispatcher = dispatcher;
        }
         public void Execute(ICommand command)
        {
            ((dynamic)this).When((dynamic)command);
        }
        public void When(CreateCustomer command)
        {
            Update(command.Id,a => a.Create(command.Id,command.CustomerName));
        }
        public void Update(CustomerId customerId, Action<CustomerAggregate> execute)
        {
            while (true)
            {
                var eventStream = _eventStore.LoadEventStream(id: customerId);
                var customer = new CustomerAggregate(eventStream.Events);
                execute(customer);
                try
                {
                     _eventStore.AppendEventsToStream(customerId, eventStream.StreamVersion, customer.Changes);
                     _dispatcher.Dispatch(customer.Changes);
                    return;
                }
                catch (OptimisticConcurrencyException ex)
                {
                    foreach (var customerEvent in customer.Changes)
                    {
                        foreach (var actualEvent in ex.ActualEvents)
                        {
                            if (ConflictsWith(customerEvent,actualEvent))
                            {
                                var msg = string.Format("Conflict between {0} and {1}",customerEvent,actualEvent);
                                throw new RealConcurrencyException(msg,ex);
                            }
                        }
                    }
                }
            }

        }
        static bool ConflictsWith(IEvent x, IEvent y)
        {
            return x.GetType() == y.GetType();
        }

            
       
    }
}
